### DISCUSSION TITLE
Unexpected token custom component

### DISCUSSION THREAD

**Discussion by eugenevk** [quasarframework/quasar#11982] [General - Components / Directives / etc] [answered]
I want to make a custom component (using the Composition API for VueJS 3) based on an existing Quasar component like QInput or QSelect, so I can preset certain characteristics for my project. I found [this example](https://v0-14.quasar-framework.org/components/building-reusable-components.html) based on QSelect and are trying to replicate it, but to no success. I also tried building a similar custom component for QInput. In both custom components I have the error of an unexpected token and I don't understand why.
Sandbox is [here](https://codesandbox.io/s/tender-chaplygin-1wkb3?file=/src/pages/InputSelect.vue).

![image](https://user-images.githubusercontent.com/19785727/148933376-3fc74a56-2514-453c-85c2-bf30453432d6.png)
![image](https://user-images.githubusercontent.com/19785727/148933455-1b53fe27-f45d-497a-b1ec-5e850c815123.png)
![image](https://user-images.githubusercontent.com/19785727/148934700-41d910e3-cb9c-404a-a5b3-ede30fbca2a8.png)

**Comment by MilosPaunovic** [comment]
`value` became `model-value` in `Vue3`

[https://quasar.dev/start/upgrade-guide#vue-3-breaking-changes-examples](https://quasar.dev/start/upgrade-guide#vue-3-breaking-changes-examples)

**Reply by eugenevk** [reply]
If I change ```value``` to ```model-value``` the error doesn't go away

**Comment by pikil** [comment]
I think, you are mixing three different issues here:
1. The one you described (does not break your component);
2. The one you have not described (doubles);
3. Another one you have not described (which actually breaks your component);

To make #2 work, rename `value` and `options` variables in `setup()` method. These are used as props in the `<template>` and that is why your component breaks - there is a double for `options` and for 'value`. Secondly, model-value instead of value as per previous suggestion as well.

To make #3 work, make sure you are not importing components as subproperties:
So, change:
```
import { InputSelect } from "./InputSelect";
import { InputText } from "./InputText";
```
to
```
import InputSelect from "./InputSelect";
import InputText from "./InputText";
```

https://codesandbox.io/s/serene-wildflower-wxbx6?file=/src/pages/Index.vue:193-270

Not sure why the editor shows unexpected token, though, but at least you can now use your component (the one in the sandbox). Check the linter options

**Reply by eugenevk** [reply]
It's weird @pikil . I can't get it to work in my app. I have copied the exact same code as in the sandbox you created into my project but get the below error message in the console:
```
runtime-core.esm-bundler.js?9e79:6620 [Vue warn]: Failed to resolve component: input-select
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement. 
  at <CompanyDetails id="" editMode="add" onVnodeUnmounted=fn<onVnodeUnmounted>  ... > 
  at <RouterView> 
  at <QPageContainer> 
  at <QLayout view="hHh lpR fFf" onAuthenticated=fn<setAuthenticatedUser> id=""  ... > 
  at <Main id="" editMode="add" onAuthenticated=fn<setAuthenticatedUser>  ... > 
  at <RouterView onAuthenticated=fn<setAuthenticatedUser> > 
  at <App>
warn @ runtime-core.esm-bundler.js?9e79:6620
```
![image](https://user-images.githubusercontent.com/19785727/149486226-fa2e6da8-3741-4a8a-8158-dd56ecd63913.png)

![image](https://user-images.githubusercontent.com/19785727/149486379-918e3965-ca40-48c3-b6e4-c34a5476b002.png)

![image](https://user-images.githubusercontent.com/19785727/149486583-ba283ed6-f147-4bf0-9c90-75fe3dcf95f1.png)

**Reply by eugenevk** [reply]
The problem must be something else. When I create and use the below simple component, I get a similar message.
```
<template>
  <div>My component</div>
</template>

<script>
export default {
  name: 'Test',
  setup () {
    return {}
  }
}
</script>
```
Any idea?

**Reply by eugenevk** [reply]
And for completeness, this is in my package.json:
```
  "dependencies": {
    "@quasar/extras": "^1.12.3",
    "algoliasearch": "^4.12.0",
    "axios": "^0.24.0",
    "core-js": "^3.6.5",
    "firebase": "^9.6.3",
    "instantsearch.css": "^7.4.5",
    "quasar": "^2.4.9",
    "vue-i18n": "^9.0.0",
    "vue-instantsearch": "^4.3.1",
    "vuex": "^4.0.1"
  },
  "devDependencies": {
    "@babel/eslint-parser": "^7.13.14",
    "@quasar/app": "^3.2.6",
    "eslint": "^7.14.0",
    "eslint-config-standard": "^16.0.2",
    "eslint-plugin-import": "^2.19.1",
    "eslint-plugin-node": "^11.0.0",
    "eslint-plugin-promise": "^5.1.0",
    "eslint-plugin-vue": "^7.0.0",
    "eslint-webpack-plugin": "^2.4.0"
  },
```

**Comment by metalsadman** [comment]
you are looking at a very old version of Quasar, wrapping components are basic for vue which is also documented in their official documentation site, I suggest you read there first.

**Comment by eugenevk** [comment, marked as answer]
I have created a new project with Quasar CLI from scratch, copying in piece by piece from my other project. For some reason now I can add a custom component without a problem. No clue why the other project was causing problems.

**Reply by eugenevk** [reply]
@metalsadman @pikil @MilosPaunovic Sorry to bother you again. Problem of "[Vue warn]: Failed to resolve component: ..." is back. I noticed the following: when I use my custom component (a country dropdown) on my Home page it works, when I use it on one of the other pages, like CompanyDetails in this case, the problem reoccurs. No idea why this happens.
The dropdown looks like this:
![image](https://user-images.githubusercontent.com/19785727/149883130-30bc4e08-e5aa-46d5-aa73-6dc494cd32b6.png)

This is my routes.js:
```
const routes = [
  { path: '/login', component: () => import('src/pages/Auth.vue') },
  {
    path: '/',
    name: 'Main',
    component: () => import('layouts/Main.vue'),
    children: [
      { path: '', name: 'Home', component: () => import('src/pages/Home.vue'), meta: { requiresAuth: true } },
      { path: 'companies', name: 'Companies', component: () => import('src/pages/Companies.vue'), meta: { requiresAuth: true } },
      { path: '/companies/create', name: 'CreateCompany', component: () => import('src/pages/CompanyDetails.vue'), props: true, meta: { requiresAuth: true } }
    ]
  },

  // Always leave this as last one, but you can also remove it
  {
    path: '/:catchAll(.*)*',
    component: () => import('pages/Error404.vue')
  }
]

export default routes
```
My Home.vue page:
```
<template>
  <q-page padding>
    ...
    <SelectCountry v-model="country" />
  </q-page>
</template>
...
<script>
import SelectCountry from 'components/SelectCountry.vue'
...
export default {
  name: 'Home',
  components: {
    SelectCountry
  },
  setup () {
    const country = ref(null)
    ...
```
My CompanyDetails.vue:
```
<template>
  <q-page padding>
    Company details
    <SelectCountry v-model="country" />
    {{ country }}
  </q-page>
</template>

<script>
import { ref } from 'vue'

import SelectCountry from 'components/SelectCountry.vue'

export default {
  name: 'CompanyDetails',
  components: [
    SelectCountry
  ],
  setup () {
    const country = ref(null)

    return {
      country
    }
  }
}
</script>
```
And this is my SelectCountry.vue:
```
<template>
  <q-select
    filter
    filter-placeholder="Select"
    label="Country"
    :model-value="value"
    :options="countries"
  >
    <template v-slot:option="scope">
      <q-item v-bind="scope.itemProps">
        <q-item-section avatar>
          <country-flag :country="scope.opt.value" size="small" />
        </q-item-section>
        <q-item-section>
          <q-item-label>{{ scope.opt.label }}</q-item-label>
        </q-item-section>
      </q-item>
    </template>
  </q-select>
</template>

<script>
import { defineComponent } from 'vue'

const countries = [
  { label: 'Afghanistan', value: 'AF', iso_3_code: 'AFG', dial: '93' },
  { label: 'Åland Islands', value: 'AX', iso_3_code: 'ALA', dial: '358' },
  ...
  { label: 'Zimbabwe', value: 'ZW', iso_3_code: 'ZWE', dial: '263' }
]

export default defineComponent({
  label: 'SelectCountry',
  props: ['value'],
  setup () {
    return {
      countries
    }
  }
})
</script>
```
I can't get this reproduced in a codepen, so hope you have a clue.

**Reply by MilosPaunovic** [reply]
Create a minimal reproduction repo and share with us.

**Reply by eugenevk** [reply]
Pfew, I found it. Very tricky one. By mistake I used the wrong brackets:
```
components: [
    SelectCountry
  ],
```
where this should be:
```
components: {
    SelectCountry
  },
```
